File: solution/752. Open the Lock/752_02.cpp
class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        unordered_set<string> q1, q2;
        unordered_set<string> dead(begin(deadends), end(deadends)); // visited
        
        string init = "0000";
        if (dead.count(init) || dead.count(target)) return -1;
        if (init == target) return 0;
        dead.insert(init);
        dead.insert(target);
        q1.insert(init);
        q2.insert(target);
        
        int level = 0;
        while (!q1.empty() && !q2.empty()) {
            if (q1.size() > q2.size())
                swap(q1, q2);
            
            unordered_set<string> q;
            for (const string& cur : q1) {
                for (int i = 0; i < 4; ++i) {
                    string next = cur;
                    for (int j = -1; j <= 1; j += 2) {
                        next[i] = (cur[i] - '0' + j + 10) % 10 + '0';
                        if (q2.count(next)) return level + 1;
                        if (dead.count(next)) continue;
                        dead.insert(next);
                        q.insert(next);
                    }
                }
            }
            swap(q1, q);
            ++level;
        }
        return -1;
    }
};

File: solution/752. Open the Lock/752_02.py
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        dead = set(deadends)
        init = "0000"
        if init in dead:
            return -1
        if init == target:
            return 0
        
        q1, q2 = {init}, {target}
        ans = 0
        while q1 and q2:
            if len(q1) > len(q2):
                q1, q2 = q2, q1
                
            q = set()
            for s in q1:
                for i in range(4):
                    for j in [-1, 1]:
                        code = s[:i] + str((int(s[i]) + j + 10) % 10) + s[i+1:]
                        if code in q2:
                            return ans + 1
                        if code in dead:
                            continue
                        dead.add(code)
                        q.add(code)
                        
            q1 = q
            ans += 1
        return -1
